README.rst
============
R10k webhook
============

Overview
--------

It's the API to `r10k utility <https://github.com/puppetlabs/r10k>`_. With it you can get your puppet code automatically deployed to puppet server after pushing to git. Assumed that you use r10k for deployment puppet environments from git branches and want to get done it automatically after pushing.

Server application works as a daemon, listens at requests and invokes local r10k utility. It's complete wrapper: generates config, launches r10k, controls and performs post-run actions. Client application is launched as VCS hook and calls server application informing about branch which has been pushed.

Features
--------

- Allows **mapping of git branches to puppet environments** using regex. E.g. branch 'master' may be mapped to puppet environment 'production'.
- Accepts **regex pattern by which name of branch is filtered**. A branch will be deployed only if matches regex.
- Runs only **one instance of r10k simultaneously. Deduplicates and keeps all requests in a queue**, so that any request won't be missed.
- **Launches 'generate types'** after r10k for `environment isolation <https://puppet.com/docs/puppet/5.5/environment_isolation.html>`_.
- Sends command **flushing cache** for an environment to api of puppet server after r10k has finished.
- Depends on only one third-party package - pyaml.

Getting started
---------------

Installation at puppet server.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Firstly, install and set `r10k utility <https://github.com/puppetlabs/r10k>`_ to your repository. Ensure that following command works.

.. code-block:: bash

    r10k deploy environment <env_name> -v

*env_name* has to be name of one of existent branches in your repo.

If you don't have errors, proceed with installation of r10k-webhook.

.. code-block:: bash

    pip3 install r10k-webhook
    systemctl daemon-reload

Enable and start service.

.. code-block:: bash

    systemctl enable r10k-webhook
    systemctl start r10k-webhook

Check logs and status.

.. code-block:: bash

    journalctl -xeu r10k-webhook
    curl localhost:8088/status

If daemon doesn't start, go on with `Configuration`_.

Installation at VCS.
~~~~~~~~~~~~~~~~~~~~

Install the package to server with your repository of puppet code.

.. code-block:: bash

    pip3 install r10k-webhook

Trigger your puppet server with <env_name>.

.. code-block:: bash

    r10k_webhook -s <puppetserver.hostname> -b <env_name>

Take <puppetserver.hostname> and <env_name> as you used at `Installation at puppet server.`_ You should see following output:

.. code-block:: bash

    Deployed the branch to 1 servers out of 1.

It means that your have deployed content of the branch to the directory of environment at puppet server host.

Use with gitolite.
^^^^^^^^^^^^^^^^^^

Create list with servers in json file looks like

.. code-block:: json

    [
      "server1",
      "server2"
    ]

Add a hook to your control repo consists of

.. code-block:: bash

    #!/usr/bin/env bash
    r10k_webhook --servers_file <path_to_servers.json>


Configuration
-------------

Create '/etc/r10k_webhook/config.json' and put there parameters.

- **host** *default: '0.0.0.0'* - Ip-address or hostname, on which http-server listening.
- **port** *default: 8088* - Port, on which http-server listening.
- **branch_to_env_map** *default: {}* - Map of name of branch in VCS and associated name of puppet environment. It may be regexp. E.g. '^env_(.\*)$': '\\g<1>' removes prefix `env_` from all branches having it.
- **allowed_branches** *default: '.\*'* - Regexp by which name of a branch is filtered. A branch will be deployed if matches regexp.
- **flush_env_cache** *default: true* - Determines whether send command flushing an environment's cache via puppet api after r10k run.
- **generate_types** *default: true* - Determines whether launch command '`puppet generate types <env> <https://puppet.com/docs/puppet/5.5/environment_isolation.html>`_' after r10k run.
- **initial_deployment** *default: true* - Deployment all environments on start.
- **override_environment_directories** *default: false* - Removes existent directory of environment before deployment.
- **r10k_path**: *default: 'r10k'* - Path to r10k binary
- **puppet_path**: *default: '/opt/puppetlabs/bin/puppet'* - Path to puppet binary
- **r10k_tmpcfg**: *default: '/tmp/r10k.yaml'* - Path to modified configuration yaml file of r10k being created and used by wrapper.
- **r10k_args**: *default: '-v'* - String with arguments are passed to r10k at every execution. Spaces are not allowed there.
- **r10k_config_path**: *default: '/etc/puppetlabs/r10k/r10k.yaml'* - Path to configuration yaml file of r10k.
- **puppet_api_uri** *default: 'https://localhost:8140/puppet-admin-api/v1'* - URI is called to flush cache of an environment.

Service r10k-webhook has to be restarted in order to apply changes of config::

    systemctl restart r10k-webhook

Example of configuration file.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: json

    {
      "flush_env_cache": false,
      "branch_to_env_map": {
        "master": "production",
        "^env_(.*)$": "\g<1>"
      },
      "allowed_branches": "^(env_[\w]+|master)$"
    }

Example of configuration file of r10k.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: yaml

    :cachedir: /opt/puppetlabs/puppet/cache/r10k
    :sources:
      puppet:
        basedir: /etc/puppetlabs/code/environments
        invalid_branches: error
        remote: git@git.example.net:puppet


CHANGELOG.rst
Changelog
=========

0.1.1 (2019-05-25)
------------------
- an environment directory may be deleted if parameter `override_environment_directories` is set.

0.1.0 (2019-05-23)
------------------
- added possibility of deploying all environments on start what controlled by parameter `initial_deployment`.
- remove limit of message size in webserver
- changed default port of puppet api

0.0.6 (2019-04-11)
------------------
- changed documentation slightly

0.0.5 (2019-04-10)
------------------
- r10k_webhook may accept branch and list of servers as cli args. Also port is parametrized.
- changed defaults: host is '0.0.0.0', flush_env_cache is True
- added documentation

0.0.4 (2019-04-04)
------------------
- Run generating environments with parameter 'codedir'.
- Remove date from logging. It's already presented by journald.
- Cleaning r10k_tmpcfg at exit.
- Removed import http.client.RemoteDisconnected from hook.
- Added feature of flushing cache via puppetserver api. But seems it's only required for old versions of r10k.

0.0.3 (2019-04-03)
------------------
- made compatible with Python 3.4
- adjusted unit file
- added parameter for path of puppet binary

0.0.2 (2019-04-02)
------------------
- added MANIFEST.in.

0.0.1 (2019-03-26)
------------------
- inited the package.


tests/test_hook.py
#!/usr/bin/env python3
from urllib.request import Request
from r10kwebhook import hook

SERVERS = [
    'fe2-t-stg-1.ae.core.sw',
    'kt-mgmt-2.starfaking.da'
]


def test_requests():
    data = '{"ref": "sampleref"}'
    servers = [hook.MgmtServer(fqdn, 8088) for fqdn in SERVERS]
    requests = [srv._get_request(data) for srv in servers]
    assert [r.full_url for r in requests] == ['http://fe2-t-stg-1.ae.core.sw:8088/api', 'http://kt-mgmt-2.starfaking.da:8088/api']
    assert [r.data for r in requests] == [data.encode(), data.encode()]


tests/test_daemon.py
#!/usr/bin/env python3
import re
import os
import pytest
import r10kwebhook


def test_is_branch_valid():
    class App(object):
        config = r10kwebhook.Settings({'allowed_branches': '^env_[a-zA-Z0-9_]+$'})
        config.allowed_branches = re.compile(config.allowed_branches)

    assert r10kwebhook.App.is_branch_valid(App, 'env_sample')
    assert not r10kwebhook.App.is_branch_valid(App, 'env_Dc-s')
    assert not r10kwebhook.App.is_branch_valid(App, 'sample')
    assert r10kwebhook.App.is_branch_valid(App, 'env_Sam_ple_')
    assert not r10kwebhook.App.is_branch_valid(App, '_env_sample')


def test_r10k_set_config(tmpdir):
    cfg_file = tmpdir.join('tmp_r10k_config.yaml')

    class R10k(object):
        config = {':cachedir': '/tmp/cache/r10k', ':sources': {
            'puppet': {'basedir': '/etc/puppetlabs/code/environments', 'invalid_branches': 'error',
                       'remote': 'dev@git.starfaking.da:puppet-dev'}}}
        _r10_cfgpath = cfg_file.realpath()
        basedirs = dict()

    r10kwebhook.R10k.set_config(R10k)
    assert cfg_file.read() == ''':cachedir: /tmp/cache/r10k
:sources:
  puppet:
    basedir: /etc/puppetlabs/code/environments.webhook
    invalid_branches: error
    remote: dev@git.starfaking.da:puppet-dev
'''
    assert R10k.basedirs == {'/etc/puppetlabs/code/environments': None}

    R10k.config = {':cachedir': '/tmp/cache/r10k', ':sources': {
        'p1': {'basedir': '/etc/puppet1', 'remote': 'dev@git.da:dev', 'prefix': '_env'},
        'p2': {'basedir': '/etc/puppet2', 'remote': 'dev@git.da:dev'},
        'p3': {'basedir': '/etc/puppet3', 'invalid_branches': 'error', 'remote': 'dev@git.da:dev', 'prefix': True}}}
    os.remove(cfg_file.realpath())
    r10kwebhook.R10k.set_config(R10k)
    assert cfg_file.read() == ''':cachedir: /tmp/cache/r10k
:sources:
  p1:
    basedir: /etc/puppet1.webhook
    prefix: _env
    remote: dev@git.da:dev
  p2:
    basedir: /etc/puppet2.webhook
    remote: dev@git.da:dev
  p3:
    basedir: /etc/puppet3.webhook
    invalid_branches: error
    prefix: true
    remote: dev@git.da:dev
'''
    assert R10k.basedirs == {'/etc/puppet1': '_env', '/etc/puppet2': None, '/etc/puppet3': 'p3'}


def test_r10k_rename_branch():
    class R10k(object):
        branch_to_env_map = {'master': 'production', '^env_(.*)$': '\g<1>'}

    assert r10kwebhook.R10k._rename_branch(R10k, 'original_name') == 'original_name'
    assert r10kwebhook.R10k._rename_branch(R10k, 'master') == 'production'
    assert r10kwebhook.R10k._rename_branch(R10k, 'pre_master', 'pre') == 'pre_production'
    assert r10kwebhook.R10k._rename_branch(R10k, 'env_original') == 'original'
    assert r10kwebhook.R10k._rename_branch(R10k, 'env_original', 'pre') == 'pre_original'
    assert r10kwebhook.R10k._rename_branch(R10k, 'pre_env_original', 'pre') == 'pre_original'


def test_metrics():
    assert r10kwebhook.App.flat_dict(
        {'requests': {'rejected': 0, 'accepted': 0}, 'r10k': {'hits': 0, 'errors': 0}}) == {
               'requests.rejected': 0, 'requests.accepted': 0, 'r10k.hits': 0, 'r10k.errors': 0}


tests/test_webserver.py
import pytest
from urllib.request import urlopen
from urllib.error import HTTPError
from r10kwebhook import webserver


def test_api():
    class Handlers(object):

        @webserver.path('/test')
        def test(self, data):
            return 'test'

    # websrv = webserver.WebServer()
    # websrv.register_handlers(Handlers())
    # try:
    #     assert urlopen('http://localhost:8088/test').read().decode() == 'test'
    #     with pytest.raises(HTTPError) as excinfo:
    #         urlopen('http://localhost:8088/other').read().decode()
    #     assert str(excinfo.value) == 'HTTP Error 404: Not Found'
    # finally:
    #     websrv.stop()


r10kwebhook/hook.py
#!/usr/bin/env python3
import json
from urllib.request import Request, urlopen
from multiprocessing.pool import ThreadPool
from urllib.error import URLError, HTTPError
from argparse import ArgumentParser, ArgumentDefaultsHelpFormatter


class MgmtServer(object):

    def __init__(self, fqdn, port):
        self.name = fqdn.split('.')[0]
        self._api_url = 'http://{}:{}/api'.format(fqdn, port)

    def _get_request(self, data):
        return Request(self._api_url, data.encode(), {'Content-Type': 'application/json'})

    def _execute(self, request):
        try:
            return urlopen(request).read().decode()
        except (URLError, HTTPError):
            return 'err'

    def deploy_ref(self, ref):
        return self._execute(self._get_request(json.dumps({'ref': ref})))


def deploy(ref, servers, port):
    pool = ThreadPool(processes=min(len(servers), 10))
    servers = [MgmtServer(fqdn, port) for fqdn in set(servers)]
    results = [pool.apply_async(srv.deploy_ref, (ref,)) for srv in servers]
    counter = [0, 0]
    for result in results:
        response = result.get()
        if response == 'ok':
            counter[0] += 1
        if response == 'wait':
            counter[1] += 1
    return counter


def main():
    parser = ArgumentParser(formatter_class=ArgumentDefaultsHelpFormatter)
    parser.add_argument('-b', '--branch', default=None, help='Branch to deploy')
    parser.add_argument('-p', '--port', default=8088, type=int, help='Port of server application')
    srvs = parser.add_mutually_exclusive_group(required=True)
    srvs.add_argument('-s', '--server', default=None, help='One or more servers', nargs='+')
    srvs.add_argument('--servers_file', default=None, help='Path to json file containing list of servers')
    args = parser.parse_args()
    if args.servers_file:
        with open(args.servers_file, 'r') as f:
            servers = json.loads(f.read())
    else:
        servers = args.server
    if args.branch:
        ref = args.branch
    else:
        ref = input().split()[2]
    deployed, triggered = deploy(ref, servers, args.port)
    if triggered:
        print('Triggered deployment of the branch at {} servers out of {}.'.format(triggered, len(servers)))
    if deployed:
        print('Deployed the branch to {} servers out of {}.'.format(deployed, len(servers)))


if __name__ == '__main__':
    main()


r10kwebhook/webserver.py
#!/usr/bin/env python3
import socket
import time
import logging
import json
from threading import Thread

logger = logging.getLogger(__name__)


def path(_path):
    def wrapper(func):
        func._path = _path
        return func

    return wrapper


class WebServer(Thread):
    """
    Class for describing simple HTTP server
    """

    def __init__(self, host='localhost', port=8088):
        Thread.__init__(self)
        self.host = host
        self.port = port
        self._stopped = False
        self._handlers = dict()
        self.start()

    def register_handlers(self, obj):
        for methodname in dir(obj):
            method = getattr(obj, methodname)
            if hasattr(method, '_path'):
                self._handlers.update({method._path: method})

    def stop(self):
        """
        Shuts down the server
        """
        logger.debug("Shutting down server")
        self._stopped = True
        try:
            self.socket.shutdown(2)
        except OSError:
            pass
        self.socket.close()

    def _generate_headers(self, response_code):
        """
        Generate HTTP response headers.
        Parameters:
            - response_code: HTTP response code to add to the header. 200 and 404 supported
        Returns:
            A formatted HTTP header for the given response_code
        """
        header = ''
        if response_code == 200:
            header += 'HTTP/1.1 200 OK\n'
        elif response_code == 404:
            header += 'HTTP/1.1 404 Not Found\n'
        elif response_code == 406:
            header += 'HTTP/1.1 406 Not Acceptable\n'
        elif response_code == 500:
            header += 'HTTP/1.1 500 Internal Server Error\n'

        time_now = time.strftime("%a, %d %b %Y %H:%M:%S", time.localtime())
        header += 'Date: {now}\n'.format(now=time_now)
        header += 'Server: r10kwebhook\n'
        header += 'Connection: close\n\n'  # Signal that connection will be closed after completing the request
        return header.encode()

    def run(self):
        """
        Attempts to create and bind a socket to launch the server. Listens on self.port for any incoming connections
        """

        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        try:
            logger.info("Starting http-server on {host}:{port}".format(host=self.host, port=self.port))
            self.socket.bind((self.host, self.port))
            logger.debug("Server started on port {port}.".format(port=self.port))

        except Exception as e:
            logger.error("Could not bind to port {port}".format(port=self.port))
            self.stop()
            return

        self.socket.listen(3)
        while not self._stopped:
            try:
                client, address = self.socket.accept()
            except (ConnectionAbortedError, OSError):
                pass
            else:
                client.settimeout(3)
                logger.debug("Recieved connection from {addr}".format(addr=address))
                Thread(target=self._handle_client, args=(client, address)).start()

    def _handle_client(self, client, address):
        """
        Main loop for handling connecting clients and serving files from content_dir
        Parameters:
            - client: socket client from accept()
            - address: socket address from accept()
        """
        PACKET_SIZE = 1024
        logger.debug("CLIENT %s", client)
        data = packet = client.recv(PACKET_SIZE)  # Recieve data packet from client
        while len(packet) == PACKET_SIZE:
            packet = client.recv(PACKET_SIZE)  # Recieve data packet from client
            data += packet
        data = data.decode()

        if not data: return

        request_method = data.split(' ')[0]
        logger.debug("Method: {m}".format(m=request_method))
        logger.debug("Request Body: {b}".format(b=data))

        if request_method in ('GET', 'POST'):
            path = data.split(' ')[1]
            if path in self._handlers:
                data = data.split('\r\n\r\n')[1]
                if data:
                    try:
                        data = json.loads(data)
                    except Exception as err:
                        logger.exception(err)
                        client.send(self._generate_headers(406))
                        client.close()
                        return
                try:
                    response_data = self._handlers[path](data).encode()
                except Exception as err:
                    logger.exception(err)
                    response = self._generate_headers(500)
                else:
                    response = self._generate_headers(200) + response_data
            else:
                response = self._generate_headers(404) + 'not found'.encode()
            client.send(response)
            client.close()
            return
        else:
            logger.debug("Unknown HTTP request method: {method}".format(method=request_method))


r10kwebhook/__init__.py
#!/usr/bin/env python3
import sys
import os
from signal import signal
import time
import logging
import subprocess
import json
import shutil
import re
from copy import deepcopy
import ssl
from urllib.request import Request, urlopen
from urllib.parse import urlencode
from threading import Lock
from argparse import ArgumentParser, ArgumentDefaultsHelpFormatter
import yaml
from r10kwebhook import webserver

logger = logging.getLogger(__name__)


class GracefulKiller(object):  # pylint: disable=too-few-public-methods
    """Catch signals to allow graceful shutdown."""

    def __init__(self):
        self.received_signal = self.received_term_signal = False
        self.last_signal = None
        caught_signals = [1, 2, 3, 10, 12, 15]
        for signum in caught_signals:
            signal(signum, self.handler)

    def handler(self, signum, frame):  # pylint: disable=unused-argument,missing-docstring
        self.last_signal = signum
        self.received_signal = True
        if signum in [2, 3, 15]:
            self.received_term_signal = True


class R10k(object):
    """ Wraps utility r10k"""

    def __new__(cls, *args):
        """ Makes singleton """
        if not hasattr(cls, 'instance'):
            cls.instance = super(R10k, cls).__new__(cls)
        return cls.instance

    def __init__(self, settings):
        self.last_run_state = 'ok'
        self.bin = settings.r10k_path
        self.generate_types = settings.generate_types
        try:
            if not self.bin or ' ' in self.bin or ';' in self.bin:  # looks like injection
                raise ValueError
            _name, self.version = subprocess.Popen([self.bin, 'version'], stdout=subprocess.PIPE, bufsize=1,
                                                   universal_newlines=True).stdout.read().strip().split()
            if _name != 'r10k':
                raise ValueError
        except (ValueError, FileNotFoundError):
            logger.error('Unable to find r10k. Set proper path to r10k binary in config, parameter \'r10k_path\'.')
            sys.exit(1)
        logger.info('Using r10k version %s', self.version)
        self.config = settings.r10k_config_path or '/etc/puppetlabs/r10k/r10k.yaml'
        if not os.path.isfile(self.config):
            logger.error('Unable to find config of r10k. Set proper path in parameter \'r10k_config_path\'.')
            sys.exit(1)
        if self.generate_types:
            self.puppet_bin = settings.puppet_path
            try:
                if not self.puppet_bin or ' ' in self.puppet_bin or ';' in self.puppet_bin:  # looks like injection
                    raise ValueError
                version = subprocess.Popen([self.puppet_bin, '-V'], stdout=subprocess.PIPE, bufsize=1,
                                           universal_newlines=True).stdout.read().strip()
            except (ValueError, FileNotFoundError):
                logger.error(
                    'Unable to find puppet. Set proper path to puppet binary in config, parameter \'puppet_path\'.')
                sys.exit(1)
            logger.info('Using puppet version %s', version)
        with open(self.config, 'r') as f:
            self.config = yaml.safe_load(f.read())
        self._r10_cfgpath = settings.r10k_tmpcfg
        if os.path.isfile(self._r10_cfgpath):
            os.remove(self._r10_cfgpath)
        self.args = settings.r10k_args.split()
        self.args.append('--config={}'.format(self._r10_cfgpath))
        self._lock = Lock()
        self._env_shelf = set()
        self.basedirs = dict()
        self.branch_to_env_map = settings.branch_to_env_map
        self.puppet_api = settings.puppet_api_uri if settings.flush_env_cache else None
        self.override_env = settings.override_environment_directories

    def set_config(self):
        if not os.path.isfile(self._r10_cfgpath):
            config = deepcopy(self.config)
            self.basedirs = dict()
            # r10k is going to put envs to temporary dirs, names of which will be appended with '.webhook'
            for source, cfg in config[':sources'].items():
                self.basedirs[cfg['basedir']] = source if cfg.get('prefix') is True else cfg.get('prefix')
                cfg['basedir'] += '.webhook'
            with open(self._r10_cfgpath, 'w') as f:
                f.write(yaml.safe_dump(config))

    def deploy_env(self, name='*'):
        if name in self._env_shelf or '*' in self._env_shelf:
            logger.warning('Requested to deploy branch %s. But it is already in queue.', name)
            return 'wait'
        self._env_shelf.add(name)
        logger.debug('Waiting for lock to deploy branch %s.', name)
        with self._lock:
            self._env_shelf.remove(name)
            logger.info('Deploying branch %s.', name)
            cmd = [self.bin, 'deploy', 'environment'] if name == '*' else [self.bin, 'deploy', 'environment', name]
            self.last_run_state = 'err' if self._exec_cmd(cmd + self.args) != 0 else 'ok'
            if self.last_run_state == 'ok':
                sync_output = self._sync_dirs()
                pack = [val for key, val in sync_output.items()] if name == '*' else [sync_output[name]]
                for basedir, env in pack:
                    if self.generate_types:  # https://puppet.com/docs/puppet/5.5/environment_isolation.html
                        logger.info('Generating types for environment \'%s\'.', env)
                        if self._exec_cmd(
                                (self.puppet_bin, 'generate', 'types', '--environment', env, '--codedir',
                                os.path.dirname(basedir))) != 0:
                            self.last_run_state = 'err'
                    if self.puppet_api:
                        logger.info('Flushing cache of environment %s. %s', env, urlopen(
                            Request('{}/environment-cache?{}'.format(self.puppet_api, urlencode({'environment': env})),
                                    method='DELETE'), context=ssl._create_unverified_context()).read().decode())
        return self.last_run_state

    def _rename_branch(self, name, prefix=None):
        """ rename prefixed branch according to setting branch_to_env_map """
        if prefix:
            name = name.replace('%s_' % prefix, '', 1)
        env = name
        for key, val in self.branch_to_env_map.items():
            if re.match(key, name):
                env = re.sub(key, val, name)
                break
        if prefix:
            return '_'.join((prefix, env))
        return env

    def _sync_dirs(self):
        """ Creates symlink from env in tmp basedir to environment in ultimate basedir. """
        renaming_map = dict()
        for basedir, prefix in self.basedirs.items():
            tmp_basedir = basedir + '.webhook'
            dir_map = dict()
            for src_dir in os.listdir(tmp_basedir):
                dir_map[src_dir] = basedir, self._rename_branch(src_dir, prefix)
            renaming_map.update(dir_map)
            os.makedirs(basedir, exist_ok=True)
            logging.debug('Cleaning symlinks which absent in r10k basedir')
            basedir_expected_content = [val[1] for val in dir_map.values()]
            for dst_dir in os.listdir(basedir):
                if dst_dir not in basedir_expected_content:
                    logger.info('Removing symlink %s', os.path.join(basedir, dst_dir))
                    os.remove(os.path.join(basedir, dst_dir))
            logger.debug('Ensuring existence of corespondent links in basedir')
            for src, dst in dir_map.items():
                if not os.path.islink(os.path.join(*dst)):
                    abs_src, abs_dst = os.path.join(tmp_basedir, src), os.path.join(*dst)
                    if self.override_env and os.path.isdir(abs_dst):
                        logger.warning('Removing directory %s', abs_dst)
                        shutil.rmtree(abs_dst)
                    logger.info('Adding symlink from %s to %s', abs_src, abs_dst)
                    os.symlink(abs_src, abs_dst, True)
        return renaming_map

    def _exec_cmd(self, args):
        """ Runs external command
        :returns exit code
        """
        self.set_config()
        logger.debug("Executing command: %s", " ".join(args))
        proc = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, bufsize=1,
                                universal_newlines=True)
        with proc.stdout:
            for line in proc.stdout:
                msg = line.split('-> ')
                if len(msg) == 2:
                    level, msg = msg
                else:
                    level, msg = 'INFO', msg[0]
                logger.log(logging.getLevelName(level.strip()), '%s -> %s', os.path.basename(args[0]), msg.strip())
        return proc.wait()


class Settings(object):
    """ Container with settings """

    def __init__(self, defaults=None):
        if isinstance(defaults, dict):
            self.__dict__.update(defaults)
        if hasattr(self, 'config_file'):
            config = self.read_config_file(self.config_file)
            if config:
                logger.info('Loading %s keys from %s', len(config), self.config_file)
                self.__dict__.update(config)
            else:
                delattr(self, 'config_file')

    @staticmethod
    def read_config_file(filepath):
        if filepath:
            if os.path.isfile(filepath):
                with open(filepath, 'r') as config_file:
                    content = json.loads(config_file.read())
                return content
            else:
                logger.warning('Cannot find config file %s', filepath)

    def __str__(self):
        return '---\n{}...'.format(json.dumps(self.__dict__, default_flow_style=False))


class App(object):

    def __init__(self):
        parser = ArgumentParser(formatter_class=ArgumentDefaultsHelpFormatter)
        parser.add_argument('-c', '--config_file', default=None, help='Path to json file containing settings')
        parser.add_argument('-rc', '--r10k_config_path', default='/etc/puppetlabs/r10k/r10k.yaml',
                            help='Path to configuration yaml file of r10k')
        parser.add_argument('-d', '--debug', action='store_true', default=False, help='Turn on debug logging')
        args = parser.parse_args()
        logging.basicConfig(level=logging.DEBUG if args.debug else logging.INFO, format='%(levelname)s: %(message)s')
        self.config = Settings({
            'config_file': args.config_file,
            'host': '0.0.0.0',
            'port': 8088,
            'r10k_path': 'r10k',
            'puppet_path': '/opt/puppetlabs/bin/puppet',
            'r10k_tmpcfg': '/tmp/r10k.yaml',
            'r10k_args': '-v',
            'r10k_config_path': args.r10k_config_path,
            'allowed_branches': '.*',
            'branch_to_env_map': dict(),
            'generate_types': True,
            'flush_env_cache': True,
            'initial_deployment': True,
            'override_environment_directories': False,
            'puppet_api_uri': 'https://localhost:8140/puppet-admin-api/v1'
        })
        self.metrics = {'requests': {'rejected': 0, 'accepted': 0}, 'r10k': {'hits': 0, 'errors': 0}}
        self._r10k = R10k(self.config)
        self._webserver = webserver.WebServer(self.config.host, self.config.port)
        self._webserver.register_handlers(self)
        if isinstance(self.config.allowed_branches, str):
            self.config.allowed_branches = re.compile(self.config.allowed_branches)
        if self.config.initial_deployment:
            if self._r10k.deploy_env() == 'err':
                self.metrics['r10k']['errors'] += 1

    @staticmethod
    def flat_dict(_dict, pkey=''):
        out = dict()
        for key, val in _dict.items():
            if isinstance(val, dict):
                out.update(App.flat_dict(val, key))
            else:
                out['.'.join((pkey, key)) if pkey else key] = val
        return out

    @webserver.path('/status')
    def check_status(self, data):
        return self._r10k.last_run_state

    @webserver.path('/metrics')
    def get_metrics(self, data):
        return '\n'.join(['='.join((k, str(v))) for k, v in self.flat_dict(self.metrics).items()])

    @webserver.path('/api')  # TODO: make REST-ful e.g. '/api/environments/<env>/deploy'
    def do(self, data):
        if 'ref' in data:
            branch = data['ref'].split('/')[-1]
            if self.is_branch_valid(branch):
                self.metrics['requests']['accepted'] += 1
                response = self._r10k.deploy_env(branch)
                if response == 'err':
                    self.metrics['r10k']['errors'] += 1
                else:
                    self.metrics['r10k']['hits'] += 1
                return response
            logger.warning('Branch name \'%s\' is invalid. Check parameter \'allowed_branches\' in config.', branch)
        self.metrics['requests']['rejected'] += 1
        return 'err'

    def run(self):
        killer = GracefulKiller()
        while not killer.received_term_signal:
            if killer.received_signal:
                logger.info("Ignoring signal %s", killer.last_signal)
                killer.received_signal = False
            if not self._webserver.is_alive():
                logger.error('Webserver is stopped. Will try to restart in 5 s')
                time.sleep(5)
                self._webserver = webserver.WebServer()
            time.sleep(0.2)
        logger.info("Received signal %s. Gracefully exiting.", killer.last_signal)
        if os.path.isfile(self.config.r10k_tmpcfg):
            logger.debug('Removing %s', self.config.r10k_tmpcfg)
            os.remove(self.config.r10k_tmpcfg)
        self._webserver.stop()

    def is_branch_valid(self, branch):
        if isinstance(self.config.allowed_branches, list) and branch in self.config.allowed_branches:
            return True
        if self.config.allowed_branches.match(branch):
            return True
        return False

    @classmethod
    def entry(cls):
        return cls().run()
